registry/remote/repository.go

```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		refDigest = d
+	}
+
+	// 4. Validate Server Digest (if present)
+	var serverHeaderDigest digest.Digest
+	if serverHeaderDigestErr := resp.Header.Get(dockerContentDigestHeader); serverHeaderDigestErr != "" {
+		if serverHeaderDigest, err = digest.Parse(serverHeaderDigestErr); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"%s %q: invalid response header value `%s`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				dockerContentDigestHeader,
+				serverHeaderDigestErr,
+				err,
+			)
+		}
 	}
 
-	size := resp.ContentLength
-	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
+	/* 5. Now, look for specific error conditions; see truth table in method docstring */
+	var contentDigest digest.Digest
 
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+	var calculatedDigest digest.Digest
+	if len(serverHeaderDigest) == 0 {
+		if httpMethod == http.MethodHead {
+			// HEAD without server `Docker-Content-Digest` header is an
+			// immediate fail; HEAD
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"HTTP %s request missing required header `%s`",
+				httpMethod, dockerContentDigestHeader,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
 
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+		// GET without server `Docker-Content-Digest` header forces the
+		// expensive calculation
+		if calculatedDigest, err = calculateDigestFromResponse(resp, maxMetadataBytes); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		contentDigest = calculatedDigest
+	} else {
+		contentDigest = serverHeaderDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if refDigest != "" && refDigest != contentDigest {
+		return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
 	}
 
+	// 6. Finally, if we made it this far, then all is good; return.
 	return ocispec.Descriptor{
 		MediaType: mediaType,
 		Digest:    contentDigest,
-		Size:      size,
+		Size:      resp.ContentLength,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response, maxMetadataBytes int64) (digest.Digest, error) {
+	defer resp.Body.Close()
+
+	body := limitReader(resp.Body, maxMetadataBytes)
```

Comment:
I think we can use the `MaxMetadataBytes` here, instead of passing the value around. As it looks like we won't set other values.

Reply:
I don't follow @Wwwsylvia, the only `MaxMetadataBytes` I can see is the declaration of the type on line 102.  Do you mean we can use the default (via `0`)?

Reply:
I think @Wwwsylvia is referring to make `calculateDigestFromResponse()` a method of `*manifestStore`.

However, the method `calculateDigestFromResponse()` looks like a common method, which can be re-used by others. Let's keep it as it is without making it a method of `*manifestStore`.

Reply:
Sorry I misunderstood something, please ignore this one.

Reply:
Lol; yeah it now makes sense to convert `generateManifestDescriptor` back to a method (`generateDescriptor`).  Done.